rocAL/source/readers/image/numpy_data_reader.cpp

```diff
@@ -0,0 +1,664 @@
+    if (!ret) {
+        parse_header(_curr_file_header, file_path);
+        if(_header_parsing_failed) {
+            ERR("Numpy header parsing failed");
+            return 0;
+        }
+        update_header_cache(file_path, _curr_file_header);
+    } else {
+        _current_file_ptr = std::fopen(file_path.c_str(), "rb");
+        if (_current_file_ptr == nullptr) {
+            ERR("Could not open file " + file_path + ": " + std::strerror(errno));
+            return 0;
+        }
+    }
+    fseek(_current_file_ptr, 0, SEEK_SET);  // Take the file pointer back to the start
+    return _curr_file_header.nbytes();
+}
+
+bool NumpyDataReader::get_header_from_cache(const std::string& file_name, NumpyHeaderData& header) {
+    std::unique_lock<std::mutex> cache_lock(_cache_mutex);
+    auto it = _header_cache.find(file_name);
+    if (it == _header_cache.end()) {
+        return false;
+    } else {
+        header = it->second;
+        return true;
+    }
+}
+
+void NumpyDataReader::update_header_cache(const std::string& file_name, const NumpyHeaderData& value) {
+    std::unique_lock<std::mutex> cache_lock(_cache_mutex);
+    _header_cache[file_name] = value;
+}
+
+const RocalTensorDataType NumpyDataReader::get_dtype(const std::string& format) {
+    if (format == "u1") return RocalTensorDataType::UINT8;
+    if (format == "u2") THROW("uint16_t dtype not supported in rocAL");
+    if (format == "u4") return RocalTensorDataType::UINT32;
+    if (format == "u8") THROW("uint64_t dtype not supported in rocAL");
+    if (format == "i1") return RocalTensorDataType::INT8;
+    if (format == "i2") return RocalTensorDataType::INT16;
+    if (format == "i4") return RocalTensorDataType::INT32;
+    if (format == "i8") THROW("int64_t dtype not supported in rocAL");
+    if (format == "f2")
+#if defined(AMD_FP16_SUPPORT)
+        return RocalTensorDataType::FP16;
+#else
+        THROW("FLOAT16 type tensor not supported")
+#endif
+    if (format == "f4") return RocalTensorDataType::FP32;
+    if (format == "f8") THROW("double dtype not supported in rocAL");
+    THROW("Unknown Numpy dtype string");
+}
+
+inline void NumpyDataReader::skip_spaces(const char*& ptr) {
+    while (::isspace(*ptr))
+        ptr++;
+}
+
+template <size_t N>
+void NumpyDataReader::skip_char(const char*& ptr, const char (&what)[N]) {
+    if (strncmp(ptr, what, N - 1)) {
+        ERR("Found wrong symbol during parsing, expected symbol: " + std::string(what));
+        _header_parsing_failed = true;
+        return;
+    }
+    ptr += N - 1;
+}
+
+template <size_t N>
+bool NumpyDataReader::try_skip_char(const char*& ptr, const char (&what)[N]) {
+    if (!strncmp(ptr, what, N - 1)) {
+        ptr += N - 1;
+        return true;
+    } else {
+        return false;
+    }
+}
+
+template <size_t N>
+void NumpyDataReader::skip_field(const char*& ptr, const char (&name)[N]) {
+    skip_spaces(ptr);
+    CALL_AND_CHECK_FLAG(skip_char(ptr, "'"));
```

Comment:
why do you need a macro here. This is inefficient. Please call the fn directly and remove the use of goto

Reply:
The macro is used to detect if the _header_parsing_failed flag is set and return abruptly instead of continuing to the next line. I have referred to this [macro](https://github.com/ROCm/MIVisionX/blob/develop/amd_openvx_extensions/amd_rpp/include/internal_rpp.h#L53) used in MIVISIONX and added a similar macro for calling function and checking flag

Reply:
I think skip_char is returning when the parsing failed with flag set. Adding another macro on top is not doing anything